math/mathcore/src/FitUtil.cxx

```diff
@@ -610,15 +610,21 @@ double FitUtil::EvaluateChi2Residual(const IModelFunction & func, const BinData
    resval = CorrectValue(resval);
 
    // estimate gradient
-   if (g != 0) {
+   if (g) {
 
       unsigned int npar = func.NPar();
-      const IGradModelFunction * gfunc = dynamic_cast<const IGradModelFunction *>( &func);
 
-      if (gfunc != 0) {
+      // use gradient of model function only if FCN support gradient
+      const IGradModelFunction * gfunc = (hasGrad) ?
+         dynamic_cast<const IGradModelFunction *>( &func) : nullptr;
+
+      if (gfunc) {
          //case function provides gradient
          if (!useBinIntegral ) {
-            gfunc->ParameterGradient(  x , p, g );
+            gfunc->ParameterGradient(  x , p, g);
+            // std::cout << "compute analytical gradient for model func at " << x[0] << " ";
+            // for (unsigned ip = 0; ip < npar; ip++) std::cout << "(" << p[ip] << ", " << g[ip] << ") " ;
+            // std::cout << std::endl;
```

Comment:
Can this commented out code be removed, or converted to some proper logging code?